### PR TITLE
Google drive document handler

### LINKED ISSUE #85: [New handler] Google Drive Documents
Add GSuite drive files into a new "documents" datastore.

Request the scope `https://www.googleapis.com/auth/drive.readonly`

Support the following document types:

- Docs (Google documents, Word documents, PDFs)
- Slides
- Spreadsheets

We don't want to store the actual document, but do want to store the text of the document and a link to the actual document.

Note: PDF files can be converted to text with this code: https://github.com/verida/data-connector-server/blob/0332d3c38b4c2c65e85a6e7fb9cb204d457b15f7/src/providers/google/helpers.ts#L133

### PR DESCRIPTION
**Pull request by chime3**
Closes #85 
### What's changed
- Get file list from google drive based on `mimeType` for `pdf` and google documents(doc, spreadsheet etc )
```
q: "mimeType='application/vnd.google-apps.document' or mimeType='application/vnd.google-apps.spreadsheet' or mimeType='application/vnd.google-apps.presentation' or mimeType='application/pdf' or mimeType='application/vnd.openxmlformats-officedocument.wordprocessingml.document'"
```
- Added `DocumentType` matching `mimeType`
- Parse `pdf` and google document only (excluding spreadsheet and presentation)

### How to test these changes
- Create at least 3 google documents and run test
`yarn run test tests/provider/google/gdrive-document.tests.ts`

### What's pending
```
export enum DocumentType {
    TXT = "txt", // Done
    PDF = "pdf", // Done
    DOC = "doc",
    DOCX = "docx",
    XLS = "xls",
    XLSX = "xlsx",
    PPT = "ppt",
    PPTX = "pptx",
    OTHER = "other" // zip, mp4????? 
}
```
- We can export google spreadsheet and presentation as `pdf` or other corresponding types like `ppt`, `xls` etc. Do you want to process them as pdf to parse?
- `doc`, `xls`, `ppt` etc are non-Google document. Do you want to parse them as well?
- There are many non-Google document in addition to `DocumentType`. Are you planning to extend supported file types? 
- Should we list non-document files in G Drive like `zip`, `mp4` as well? I mean `OTHER` type will include `zip`, `mp4` etc?

### CODE REVIEW

**Comment by tahpot** [comment]
We don't want to store the files, we are just extracting text from documents that have text content.

>We can export google spreadsheet and presentation as pdf or other corresponding types like ppt, xls etc. Do you want to process them as pdf to parse?

Let's ignore spreadsheets. Convert presentations to PDF, then parse them.

>doc, xls, ppt etc are non-Google document. Do you want to parse them as well?

Yes to `.doc` and `.ppt`.

Instead of downloading the files and parsing them, is it possible to fetch the `indexableText` for each file? See https://developers.google.com/drive/api/guides/file#indexable-text

According to ChatGPT:

>To fetch the indexable text property of a Google Drive file using the Google Drive API, you'll need to use the files.get method with the fields parameter specifying the indexableText field.

>There are many non-Google document in addition to DocumentType. Are you planning to extend supported file types?

Not at this stage.

>Should we list non-document files in G Drive like zip, mp4 as well? I mean OTHER type will include zip, mp4 etc?

I think we should list them and include a link to them, but not parse them.

**Review by tahpot — CHANGES_REQUESTED** [review]
Changes requested as per comment above

**Comment by chime3** [comment]
> Instead of downloading the files and parsing them, is it possible to fetch the `indexableText` for each file? See https://developers.google.com/drive/api/guides/file#indexable-text

Extracted `indexableText` from google documents/slides/presentations (PUSHED)

In terms of non-Google document, to avoid downloading we can convert 
- PDF/DOC/DOCX => google document
- PPT/PPTX => google presentation
- XLS/XLSX => google spreadsheet

and then can get `indexableText` from converted documents.
But this requires `write` permission.
```
      // Convert the file(doc/docx, ppt/pptx, xls/xlsx) to a Google Docs/Slides/Presentations format
      const response = await drive.files.copy({
        fileId: fileId,
        requestBody: {
          mimeType: mimeType, // Convert to Google Docs/Slides/Presentations
        },
      });

      const documentId = response.data.id; // newly created file ID
      
      // Process as like google documents      
```
Otherwise, we should download non-Google document to parse content.

Please confirm which one we will proceed, @tahpot

**Comment by chime3** [comment]
> We don't want to store the files, we are just extracting text from documents that have text content.

Yeah, we are extracting text now not storing file.

**Comment by chime3** [comment]
> Let's not parse spreadsheets. Convert presentations to PDF, then parse them.
We can get `indexableText` from google spreadsheets now. Do you still not want to parse spreadsheets?

**Comment by chime3** [comment]
> Yes to parsing `.doc` and `.ppt`.
Okay
> Not at this stage.
Okay
> I think we should list them and include a link to them, but not parse them.
Okay

**Comment by tahpot** [comment]
> We can get `indexableText` from google spreadsheets now. Do you still not want to parse spreadsheets?

Let's use `indexableText` for any file that supports it, so yes to spreadsheets.

**Comment by tahpot** [comment]
> > Instead of downloading the files and parsing them, is it possible to fetch the `indexableText` for each file? See https://developers.google.com/drive/api/guides/file#indexable-text
> 
> Extracted `indexableText` from google documents/slides/presentations (PUSHED)
> 
> In terms of non-Google document, to avoid downloading we can convert
> 
> * PDF/DOC/DOCX => google document
> * PPT/PPTX => google presentation
> * XLS/XLSX => google spreadsheet
> 
> and then can get `indexableText` from converted documents. But this requires `write` permission.

Let's download and parse. Add an option in the handler for `maxFileSizeToIndex` with default = 10MB.

File: src/providers/google/gdrive-document.ts [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+    public getSchemaUri(): string {
+        return CONFIG.verida.schemas.DOCUMENT;
+    }
+
+    public getProviderApplicationUrl(): string {
+        return "https://drive.google.com";
+    }
+    
+    public getGoogleAuth(): OAuth2Client {
```
Comment (tahpot) [inline src/providers/google/gdrive-document.ts:31]: Inherit from GoogleHandler and use parent `getGoogleAuth()` method.

File: src/providers/google/gdrive-document.ts [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+    public getGoogleDrive(): drive_v3.Drive {
+        
+        const auth = this.getGoogleAuth();
+
+        const drive = google.drive({ version: "v3", auth: auth });
+        return drive;
+    }
+
+    public async _sync(
```
Comment (tahpot) [inline src/providers/google/gdrive-document.ts:60]: Update `_sync()` to use the new `ItemsRangeTracker` strategy.

File: src/providers/google/gdrive-document.ts [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+        if (syncPosition.thisRef) {
+            query.pageToken = syncPosition.thisRef;
+        }
+
+        const serverResponse = await drive.files.list(query);
+
+        if (
+            !_.has(serverResponse, "data.files") ||
+            !serverResponse.data.files.length
```
Comment (tahpot) [inline src/providers/google/gdrive-document.ts:81]: `data.files` is possibly undefined

File: src/providers/google/gdrive-document.ts [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        const results: SchemaDocument[] = [];
+        for (const file of serverResponse.data.files) {
+            const fileId = file.id;
+
+            if (fileId == breakId) {
+                const logEvent: SyncProviderLogEvent = {
+                    level: SyncProviderLogLevel.DEBUG,
+                    message: `Break ID hit (${breakId})`
+                }
+                this.emit('log', logEvent)
```
Comment (tahpot) [inline src/providers/google/gdrive-document.ts:166]: Can't set as "Unknown" as it's not an ISO Date/time. Set it to now.

File: src/providers/google/gdrive-document.ts [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+            const modifiedTime = file.modifiedTime || "Unkown";
+            
+            if (breakTimestamp && modifiedTime < breakTimestamp) {
+                const logEvent: SyncProviderLogEvent = {
+                    level: SyncProviderLogLevel.DEBUG,
+                    message: `Break timestamp hit (${breakTimestamp})`
+                }
+                this.emit('log', logEvent)
+                break;
```
Comment (tahpot) [inline src/providers/google/gdrive-document.ts:178]: If there is no link, ignore this file.

When would this possibly happen? I thought all gdrive files have links.

File: src/providers/google/gdrive-document.ts [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+            }
+
```
Comment (tahpot) [inline src/providers/google/gdrive-document.ts:180]: If there is no mimeType, let's ignore this file.

File: src/providers/google/gdrive-document.ts [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+            const title = file.name || "No title";
```
Comment (tahpot) [inline src/providers/google/gdrive-document.ts:181]: "No thumbnail" is not a valid icon, so you can't set it to that. Set as `undefined`.

File: src/providers/google/helpers.ts [hunk 8]
```diff
@@ -171,3 +173,245 @@ export class GmailHelpers {
...
+        { fileId: fileId, mimeType: "text/plain" }, 
+        { responseType: "arraybuffer" }
+      );
+      return Buffer.byteLength(exportedFile.data as ArrayBuffer);
+    } else {
+      return undefined;
+    }
+  }
+
+  static async extractIndexableText(
```
Comment (tahpot) [inline src/providers/google/helpers.ts:215]: What happened to using `indexableText`?

We don't want to download files if we don't have to.

**Review by tahpot — CHANGES_REQUESTED** [review]

File: src/providers/google/gdrive-document.ts [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+            const textContent = await GoogleDriveHelpers.extractIndexableText(drive, file.id, mimeType, auth);
+            
+            results.push({
+                _id: this.buildItemId(fileId),
+                name: title,
+                type: type,
+                size: size,
+                uri: link,
+                icon: thumbnail,
+                contentText: textContent,
```
Comment (tahpot) [inline src/providers/google/gdrive-document.ts:196]: You can also set `modifiedAt` (see base schema: https://core.schemas.verida.io/base/v0.1.0/schema.json)

File: src/providers/google/gdrive-document.ts [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+                sourceId: file.id,
```
Comment (tahpot) [inline src/providers/google/gdrive-document.ts:197]: Does google provide `inserted` timestamp that is different from `modified`?

File: src/providers/google/gdrive-document.ts (hunk 1 above)
Comment (chime3) [reply src/providers/google/gdrive-document.ts:31]: > Inherit from GoogleHandler and use parent `getGoogleAuth()` method.

Solved

File: src/providers/google/gdrive-document.ts (hunk 2 above)
Comment (chime3) [reply src/providers/google/gdrive-document.ts:60]: > Update `_sync()` to use the new `ItemsRangeTracker` strategy.

Done

File: src/providers/google/gdrive-document.ts (hunk 3 above)
Comment (chime3) [reply src/providers/google/gdrive-document.ts:81]: > `data.files` is possibly undefined

changed

File: src/providers/google/gdrive-document.ts (hunk 4 above)
Comment (chime3) [reply src/providers/google/gdrive-document.ts:166]: > Can't set as "Unknown" as it's not an ISO Date/time. Set it to now.

Fixed

File: src/providers/google/gdrive-document.ts (hunk 5 above)
Comment (chime3) [reply src/providers/google/gdrive-document.ts:178]: > If there is no link, ignore this file.
> 
> When would this possibly happen? I thought all gdrive files have links.

In-completed/draft files and restricted files don't have `webviewlink`, and ignored.

File: src/providers/google/gdrive-document.ts (hunk 6 above)
Comment (chime3) [reply src/providers/google/gdrive-document.ts:180]: > If there is no mimeType, let's ignore this file.

Done

File: src/providers/google/gdrive-document.ts (hunk 7 above)
Comment (chime3) [reply src/providers/google/gdrive-document.ts:181]: > "No thumbnail" is not a valid icon, so you can't set it to that. Set as `undefined`.

Done

File: src/providers/google/gdrive-document.ts (hunk 10 above)
Comment (chime3) [reply src/providers/google/gdrive-document.ts:197]: > Does google provide `inserted` timestamp that is different from `modified`?

Added `createdTime`

File: src/providers/google/gdrive-document.ts (hunk 9 above)
Comment (chime3) [reply src/providers/google/gdrive-document.ts:196]: > You can also set `modifiedAt` (see base schema: https://core.schemas.verida.io/base/v0.1.0/schema.json)

Done

File: src/providers/google/helpers.ts (hunk 8 above)
Comment (chime3) [reply src/providers/google/helpers.ts:215]: > What happened to using `indexableText`?
> 
> We don't want to download files if we don't have to.

Doing some tests

**Comment by tahpot** [comment]
One more change.

Please switch from using the document schema, to using this new file schema: https://common.schemas.verida.io/file/v0.1.0/schema.json

For google documents, `fileDataId` should always be undefined, but `uri` should always be defined.

**Comment by chime3** [comment]
> One more change.
> 
> Please switch from using the document schema, to using this new file schema: https://common.schemas.verida.io/file/v0.1.0/schema.json
> 
> For google documents, `fileDataId` should always be undefined, but `uri` should always be defined.

Updated schema and passed unit test. Also checked Google drive data using web interface.
gdrive-tests
image

File: src/providers/google/helpers.ts (hunk 8 above)
Comment (chime3) [reply src/providers/google/helpers.ts]: > What happened to using `indexableText`?
> 
> We don't want to download files if we don't have to.

I've added extracting `indexableText` if available. Very rare to check indexable text and it takes some time to index documents for GDrive

**Comment by chime3** [comment]
One more thing in extracting extension @tahpot 
- Extracting extension from file name using period first: this can cause error because filenames might use periods for its own, not for extension.
- If not possible from file name, then use `mimetype` to determine extension.

We can not recognize all file extensions.
Please clarify on this if we use type constraints or not.

File: src/web/developer/data/data.js [hunk 11]
```diff
@@ -209,6 +209,7 @@ $(document).ready(function() {
             "Social Post": "https://common.schemas.verida.io/social/post/v0.1.0/schema.json",
             "Favourites": "https://common.schemas.verida.io/favourite/v0.1.0/schema.json",
             "Email": "https://common.schemas.verida.io/social/email/v0.1.0/schema.json",
+            "FILE": "https://common.schemas.verida.io/file/v0.1.0/schema.json",
```
Comment (tahpot) [inline src/web/developer/data/data.js:212]: Just `File`

File: src/providers/google/helpers.ts [hunk 12]
```diff
@@ -171,3 +265,296 @@ export class GmailHelpers {
...
+
+      // Check if indexable text is available
+      textContent = response.data.contentHints?.indexableText || "";
+
+      // If no indexable text is found, proceed with original extraction methods
+      if (!textContent) {
+        console.warn("No indexable text found, using fallback method.");
+
+        // 10MB limit (5 * 1024 * 1024)
+        const sizeLimit = 10 * 1024 * 1024;
```
Comment (tahpot) [inline src/providers/google/helpers.ts:330]: Set this as a config option in the google config.

**Comment by tahpot** [comment]
Feedback:

- Can confirm tests pass
- There was an issue with Google Slides API not being enabled, so I have enabled it. Will we have the same issue with other types like drawings or sheets?
- When running the sync on my gmail, I receive this error in the sync logs and no files are saved (`Export only supports Docs Editors files.`). From [this stackoverflow](https://stackoverflow.com/questions/65053558/google-drive-api-v3-files-export-method-throws-a-403-error-export-only-support) it indicates `files.get()` may need to be used instead?
- The failure of one item, shouldn't break the whole sync process, so I'm not sure what is happening to prevent any items being saved.

**Review by tahpot — CHANGES_REQUESTED** [review]
See comments above

File: src/web/developer/data/data.js (hunk 11 above)
Comment (chime3) [reply src/web/developer/data/data.js]: > Just `File`

Fixed

File: src/providers/google/gdrive-document.ts [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+        drive: drive_v3.Drive,
+        serverResponse: GaxiosResponse<drive_v3.Schema$FileList>,
+        breakId: string,
+        breakTimestamp?: string
+    ): Promise<SyncDocumentItemsResult> {
+        const results: SchemaDocument[] = [];
+        let breakHit: SyncItemsBreak;
+
+        for (const file of serverResponse.data.files) {
+            const fileId = file.id;
```
[the remark this answers is not in the document]
Comment (chime3) [reply src/providers/google/gdrive-document.ts]: > According to typescript `file.id` could be null. Check for that and break if it is null.

Fixed

File: src/providers/google/helpers.ts (hunk 12 above)
Comment (chime3) [reply src/providers/google/helpers.ts:330]: > Set this as a config option in the google config.

Done

**Comment by chime3** [comment]
### What has been changed?
- Replace `pptx-parser` with `officeparser`
- Ignored `doc` and `ppt` because those types are deprecated in recent libraries and we need to use `libreoffice-convert` to convert `doc/ppt` into `docx/pptx` to use trending packages, but `libreoffice-convert` requires libreoffice pre installed. So ignored parsing `doc` and `ppt`.

**Comment by chime3** [comment]
> Feedback:
> 
> * Can confirm tests pass
> * There was an issue with Google Slides API not being enabled, so I have enabled it. Will we have the same issue with other types like drawings or sheets?
> * When running the sync on my gmail, I receive this error in the sync logs and no files are saved (`Export only supports Docs Editors files.`). From [this stackoverflow](https://stackoverflow.com/questions/65053558/google-drive-api-v3-files-export-method-throws-a-403-error-export-only-support) it indicates `files.get()` may need to be used instead?
> * The failure of one item, shouldn't break the whole sync process, so I'm not sure what is happening to prevent any items being saved.

All done

**Review by tahpot — COMMENTED** [review]
All good to merge once the handler is updated to handle the last page fix in the other google handlers.

**Comment by chime3** [comment]
> All good to merge once the handler is updated to handle the last page fix in the other google handlers.

Okay, thank you

**Pull request merged**